include/pipe.h

```diff
@@ -1,153 +1,42 @@
 
-#include <cmath>
-
-#include <array>
-#include <exception>
-#include <memory>
-
-#include "node.h"
-#include "settings.h"
+#ifndef PIPE_NETWORK_PIPE_H
+#define PIPE_NETWORK_PIPE_H
+#include "link.h"
 
 namespace pipenetwork {
 
-//! Pipe class
-//! \brief Class that stores the information about pipes
-class Pipe {
-
+class Pipe : public Link {
  public:
-  //! Constructor with pipe id, node pointers, diameter, status and max
-  //! allowable velocity
-  //! \param[in] id pipe id
-  //! \param[in] nodes array of node pointers
-  //! \param[in] diameter of the pipe
-  //! \param[in] roughness pipe roughness for Hazen-Williams equation
-  //! \param[in] status of the pipe, true for open, flase for close
-  //! \param[in] maximum allowable velocity
-  Pipe(unsigned id,
-       const std::array<std::shared_ptr<pipenetwork::Node>, 2>& nodes,
-       double diameter, double roughness, bool status,
-       double max_velocity = std::numeric_limits<double>::max());
-
-  //! Destructor
-  ~Pipe() { nodes_.fill(nullptr); }
-
-  //! Copy constructor
-  Pipe(const Pipe&) = delete;
-
-  //! Assignment operator
-  Pipe& operator=(const Pipe&) = delete;
-
-  //! Move constructor
-  Pipe(Pipe&&) = delete;
-
-  //! Return id
-  //! \retval id_ id of the pipe
-  unsigned id() const { return id_; }
-
-  //! Return length of the pipe
-  //! \retval length_ pipe length
-  double length() { return length_; }
-
-  //! Assign Dracy friction factor
-  //! \param[in] darcy_friction Darcy friction factor of the pipe
-  void darcy_friction(double darcy_friction) {
-    darcy_friction_ = darcy_friction;
-  }
-
-  //! Assign discharge to the pipe
-  //! \param[in] iter_discharge dischargeof the pipe during iteration
-  void iter_discharge(double iter_discharge) {
-    iter_discharge_ = iter_discharge;
-  }
-
-  //! Initialize discharge with input value
-  //! \param[in] discharge input discharge value of the pipe
-  void initialize_discharge(double discharge = 0.001) {
-    discharge_ = discharge;
-    iter_discharge_ = discharge_;
+  //! Constructor with two end nodes, length, diameter roughness and pipe status
+  //! \param[in] node1 ptr for end node
+  //! \param[in] node2 ptr for end node
+  //! \param[in] length, length of the pipe
+  //! \param[in] diameter, diameter of the pipe
+  //! \param[in] roughness, roughness of the pipe
+  //! \param[in] status, status of the pipe (open or close)
+  Pipe(Index id, const std::shared_ptr<pipenetwork::Node>& node1,
+       const std::shared_ptr<pipenetwork::Node>& node2, const double length,
+       const double diameter, const double roughness, const Pipe_status status)
+      : Link(id, node1, node2) {
+    pipe_info_["type"] = PIPE;
+    pipe_info_["length"] = length;
+    pipe_info_["diameter"] = diameter;
+    pipe_info_["roughness"] = roughness;
+    pipe_info_["status"] = status;
+  };
+
+  //! Virtual destructor
+  ~Pipe() override{};
+
+  //! Return link info
+  std::map<std::string, double> link_info() const override {
```

Comment:
what's the key?

Reply:
Keys are defined in the constructor above